v0.0.5

sol-znn opened the pull request:
## v0.0.5

This branch adds support for the following:
- Refactored to align with the [znn_cli_csharp](https://github.com/hypercore-one/znn_cli_csharp)
- Added support for HTLC, Spork, Bridge, Liquidity contracts
- Added admin-only functionality (Bridge/Liquidity)
- Added ability to donate to AZ and query Stats RPC
- Added `chainId` option (default, auto, manual)
- Updated menu, including an admin-only menu
   - Updated `verbose` flag to display a brief description for commands
   - Updated `help` flag to display all options for a category
- Updated Makefile: bugfix + support for Windows

**Limitations**
- Some [orchestrator-specific](https://github.com/hypercore-one/znn_cli_dart/blob/feature/v0.0.5/lib/orchestrator.dart) functions were not included due to low value/increased complexity.  
These can be omitted since they're handled by orchestrators.

sol-znn committed:
v0.0.5

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+void generalMenu() {
+  print('  ${white('General')}');
+  print(
+      '    send toAddress amount [${green('ZNN')}/${blue('QSR')}/${magenta('ZTS')}]');
```
The optional message argument is missing.

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+  print('    unconfirmed');
+  print('    balance address');
+  print('    frontierMomentum');
+  print('    createHash "string" [hashType preimageLength]');
```
`createHash` only has two arguments `[hashType preimageLength]`

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+
+void _version() {
+  print('$znnCli v$znnCliVersion using Zenon SDK v$znnSdkVersion');
+  print(getZnndVersion(znnDaemon));
```
Why show the znnd information?

sol-znn replied:
Agree, it wasn't really valuable information.

KingGorrin replied:
The `process.info` command shows all the info you need.

Also, we need to do something about the versioning schema. I would like to align my versions, but I'm using semantic versioning.

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  BigInt amount =
+      AmountUtils.extractDecimals(num.parse(args[2]), token.decimals);
+
```
Negative or zero amounts are not checked.

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+  var block = AccountBlockTemplate.send(recipient, tokenStandard, amount);
+
+  if (args.length == 5) {
+    block.data = AsciiEncoder().convert(args[4]);
```
Can UFT8 encoding be used instead?

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+  switch (hashType) {
+    case 1:
+      hash = Hash.fromBytes(await Crypto.sha256Bytes(preimage));
+      print('SHA-256 Hash: $hash');
```
SHA2-256

KingGorrin commented on src/general.dart:
```diff
@@ -0,0 +1,326 @@
+      return;
+    default:
+      hash = Hash.digest(preimage);
+      print('SHA-3 Hash: $hash');
```
SHA3-256

KingGorrin commented on src/htlc.dart:
```diff
@@ -0,0 +1,657 @@
+  int htlcTimelockMaxHours = htlcTimelockMinHours * 24;
+
+  try {
+    hashLockedAddress = Address.parse(args[1]);
```
No checks are made to prevent empty or embedded addresses.

KingGorrin commented on src/htlc.dart:
```diff
@@ -0,0 +1,657 @@
+  }
+
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  amount = AmountUtils.extractDecimals(num.parse(args[3]), token.decimals);
```
Amount as string?

KingGorrin replied:
Forget this one.

KingGorrin commented on src/htlc.dart:
```diff
@@ -0,0 +1,657 @@
+    switch (hashType) {
+      case 1:
+        hashLock = Hash.fromBytes(await Crypto.sha256Bytes(preimage));
+        return;
```
I think this needs to be a `break`?

KingGorrin commented on src/htlc.dart:
```diff
@@ -0,0 +1,657 @@
+    print('${red('Error!')} amount must be greater than 0');
+    return;
+  }
+
```
No balance check is made. `if (!(await hasBalance()) return;`

KingGorrin commented on src/htlc.dart:
```diff
@@ -0,0 +1,657 @@
+    return;
+  }
+
+  Hash blockHash = Hash.parse(args[1]);
```
catch Hash.parse exception.

KingGorrin commented on src/htlc.dart:
```diff
@@ -0,0 +1,657 @@
+
+  switch (hashType) {
+    case 1:
+      print('HashType 1 detected. Encoding preimage to SHA-256...');
```
SHA2-256

KingGorrin commented on src/accelerator.dart:
```diff
@@ -0,0 +1,56 @@
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  BigInt amount =
+      AmountUtils.extractDecimals(num.parse(args[1]), token.decimals);
+  if (amount == BigInt.zero) {
```
Check if amount is bigger than 0. `if (amount <= BigInt.zero)`

KingGorrin commented on src/accelerator.dart:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  print(
+      'Donating ${AmountUtils.addDecimals(amount, token.decimals)} ${token.symbol} to Accelerator Z ...');
```
Accelerator-Z

KingGorrin commented on src/pillar.dart:
```diff
@@ -0,0 +1,224 @@
+  print('Registering Pillar ...');
+  await znnClient.send(znnClient.embedded.pillar.register(
+      newName,
+      Address.parse(args[2]),
```
Parse and check empty or embedded.

KingGorrin commented on src/plasma.dart:
```diff
@@ -0,0 +1,154 @@
+    print('plasma.fuse toAddress amount (in ${blue('QSR')})');
+    return;
+  }
+  Address beneficiary = Address.parse(args[1]);
```
Check parse and empty or embedded address.

KingGorrin commented on src/plasma.dart:
```diff
@@ -0,0 +1,154 @@
+    print('plasma.cancel id');
+    return;
+  }
+  Hash id = Hash.parse(args[1]);
```
Try/catch parse

KingGorrin commented on src/staking.dart:
```diff
@@ -0,0 +1,155 @@
+    print('stake.revoke id');
+    return;
+  }
+  Hash hash = Hash.parse(args[1]);
```
Try/catch hash

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+    print('token.getByStandard tokenStandard');
+    return;
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Use GetTokenStandard()

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+    return;
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
```
Check if token exist

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+    return;
+  }
+  String type = 'Token';
+  Address ownerAddress = Address.parse(args[1]);
```
Check parse address.

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+    return;
+  }
+  print('Transferring ZTS token ownership ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Use GetTokenStandard()

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+    print('token.mint tokenStandard amount receiveAddress');
+    return;
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Use GetTokenStandard()

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+  }
+  print('Transferring ZTS token ownership ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  Address newOwnerAddress = Address.parse(args[2]);
```
Check parse address

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+    return;
+  }
+  print('Disabling ZTS token mintable flag ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Use GetTokenStandard()

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+  }
+  print('Disabling ZTS token mintable flag ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  var token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
```
Check if token exists

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+  print('Transferring ZTS token ownership ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  Address newOwnerAddress = Address.parse(args[2]);
+  var token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
```
Check if token exists

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+    return;
+  }
+
+  Address address = Address.parse(args[1]);
```
Try/Catch parse address

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+    return;
+  }
+
+  Address address = Address.parse(args[1]);
```
Try/catch parse address

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+    return;
+  }
+
+  Address address = Address.parse(args[1]);
```
Try/catch parse address

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+
+  int months = int.parse(args[1]);
+  int duration = months * 24 * 60 * 60;
+  TokenStandard tokenStandard = TokenStandard.parse(args[3]);
```
Use GetTokenStandard

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+    return;
+  }
+
+  Hash id = Hash.parse(args[1]);
```
Try/catch parse hash

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+  String currentAdmin = (await znnClient.embedded.liquidity.getLiquidityInfo())
+      .administrator
+      .toString();
+  Address newAdmin = Address.parse(args[1]);
```
Try/catch parse address and check new admin must be an user address.

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+    return;
+  }
+
+  Address newAdmin = Address.parse(args[1]);
```
Try/catch parse address and check not empty or embedded

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+      return;
+    }
+  }
+
```
Nomination works with time challenges, see  C# impl. for details.

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+  for (int i = 1; i < args.length; i++) {
+    try {
+      Address guardian = Address.parse(args[i]);
+      if (guardian != emptyAddress) {
```
Check not empty or embedded and all are distinct

KingGorrin commented on src/liquidity.dart:
```diff
@@ -0,0 +1,625 @@
+Future<void> _unlockStakeEntries() async {
+  AccountBlockTemplate unlockLiquidityStakeEntries =
+      znnClient.embedded.liquidity.unlockLiquidityStakeEntries(
+          TokenStandard.parse('zts17d6yr02kh0r9qr566p7tg6'));
```
Why is the token standard hardcoded?

KingGorrin commented on src/plasma.dart:
```diff
@@ -0,0 +1,154 @@
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Assert pageIndex/pageSize constraints

KingGorrin commented on src/spork.dart:
```diff
@@ -0,0 +1,106 @@
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Assert pageIndex/pageSize constraints

KingGorrin commented on src/staking.dart:
```diff
@@ -0,0 +1,155 @@
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Assert pageIdex/pageSize constraints

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Assert pageIndex/pageSize constraints

KingGorrin commented on src/token.dart:
```diff
@@ -0,0 +1,353 @@
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  BigInt amount = BigInt.parse(args[2]);
+  Address mintAddress = Address.parse(args[3]);
```
Try/catch parse address

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+  BigInt amount =
+      AmountUtils.extractDecimals(num.parse(args[4]), token.decimals);
+
+  if (amount == BigInt.zero) {
```
amount must be greater than 0

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+    return;
+  }
+
+  Hash id = Hash.parse(args[1]);
```
Try/catch parse hash

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+    return;
+  }
+
+  Hash transactionHash = Hash.parse(args[1]);
```
Try/catch parse hash

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+  print('   Log Index: ${request.logIndex}');
+  print('   To: ${request.toAddress}');
+  print('   Token Standard: ${request.tokenStandard}');
+  print('   Amount: ${AmountUtils.addDecimals(request.amount, decimals)}');
```
Token symbol is not displayed

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+  print(
+      '   From: ${(await znnClient.ledger.getAccountBlockByHash(request.id))?.address}');
+  print('   Token Standard: ${request.tokenStandard}');
+  print('   Amount: ${AmountUtils.addDecimals(request.amount, decimals)}');
```
Token symbol is not displayed

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+  String currentAdmin = (await znnClient.embedded.bridge.getBridgeInfo())
+      .administrator
+      .toString();
+  Address newAdmin = Address.parse(args[1]);
```
Check parse and assert not empty or embedded address

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+Future<void> _halt() async {
+  _isAdmin();
+  print('Halting bridge ...');
+  AccountBlockTemplate halt = znnClient.embedded.bridge.halt('');
```
Use signature value '1' to circumvent the empty string unpack issue.

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+  Hash transactionHash = Hash.parse(args[1]);
+  int logIndex = int.parse(args[2]);
+
+  print('Removing token pair ...');
```
Invalid print statement.

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+      return;
+    }
+  }
+
```
Nomination works with time challenges, see C# impl. for details.

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+    return;
+  }
+
+  Address newAdmin = Address.parse(args[1]);
```
Check parse and assert not empty or embedded address.

KingGorrin commented on src/bridge.dart:
```diff
@@ -0,0 +1,1197 @@
+      '    bridge.admin.setNetwork networkClass chainId name contractAddress metadata');
+  print('    bridge.admin.removeNetwork networkClass chainId');
+  print('    bridge.admin.setNetworkMetadata networkClass chainId metadata');
+}
```
`bridge.admin.setRedeemDelay` is missing?

sol-znn replied:
Omitted on purpose due to incomplete implementation in go-zenon.

1. [Definition statement](https://github.com/zenon-network/go-zenon/blob/master/vm/embedded/definition/bridge.go#L101)
2. [Missing abi](https://github.com/zenon-network/go-zenon/blob/master/vm/embedded/embedded.go#L59)
3. [Missing SetRedeemDelayMethod](https://github.com/zenon-network/go-zenon/blob/master/vm/embedded/implementation/bridge.go)

KingGorrin replied:
I should have checked go-zenon. Thanks for the insight.

KingGorrin reviewed (changes requested):
@Sol-Sanctum I've reviewed the changes and made lots of inline comments. Most are about assertions and inconsistencies between other methods. I'm not sure about all the comments, so if you can give your rationale I'd appreciate it. I've already included your changes in the `bigint` branch of the C# CLI. I await your answers in order to finalize the implementation.

sol-znn committed:
Finished refactoring, added input validation and error handling

sol-znn commented:
Thank you for reviewing my work and providing detailed feedback. 
I believe I've addressed most of your comments in this commit: https://github.com/hypercore-one/znn_cli_dart/pull/2/commits/77a3b7fae06f1e713d4c0027ed169d70c4951908

A few stood out, though:
1. [Semantic versioning](https://github.com/hypercore-one/znn_cli_dart/pull/2#discussion_r1263384647)
2. I don't understand [this one](https://github.com/hypercore-one/znn_cli_dart/pull/2#discussion_r1259528074)?
3. I don't think [this](https://github.com/hypercore-one/znn_cli_dart/pull/2#discussion_r1260966259) needs a populated string but we can fix it in a future commit. Will have to check out [this issue.](https://github.com/hypercore-one/znn_sdk_dart/issues/1)

KingGorrin commented on lib/accelerator.dart:
```diff
@@ -1,6 +1,6 @@
-import 'src.dart';
 
 void acceleratorMenu() {
   print('  ${white('Accelerator Z')}');
```
Accelerator-Z 🤣

KingGorrin commented:
Very nice work in total!

sol-znn committed:
String update

sol-znn committed:
Updated BigInt formatting to remove 'num' type

sol-znn committed:
Updated BigInt formatting to remove 'num' type

sol-znn committed:
Refactor amount formatting

sol-znn committed:
Revert "Refactor amount formatting"

This reverts commit 6215b6c1fc844d85d7ae5356fe10885e478ac02d.

sol-znn committed:
Reverted AmountUtils removal

sol-znn committed:
Reverted AmountUtils removal

sol-znn committed:
Added support for liquidity.setTokenTuple()

sol-znn committed:
chore: merged master with branch

sol-znn committed:
Simplified some variables

sol-znn committed:
Fix: invalid plasma.cancel return statement

sol-znn committed:
Merge pull request #3 from hypercore-one/fix/cancel_fuse

Fix: invalid plasma.cancel return statement

sol-znn committed:
Updated permission for wallet.export

sol-znn committed:
Fixed unwrap print statement
